### ISSUE TITLE
CancelScope shielding fails to protect against cancellation until task_status.started is called on asyncio

### ISSUE THREAD

**Issue by tapetersen** [agronholm/anyio#837]
### Things to check first

- [X] I have searched the existing issues and didn't find my bug already reported there

- [X] I have checked that my bug is still present in the latest release

### AnyIO version

4.7.0

### Python version

3.10

### What happened?

Running some quite involved code that needs to do shielded cleanup if it fails sometime exits a shielded scope due to external cancellation if it was cancelled before it had called `task_status.started()` 

I know these are edge cases and may also be really hard to solve so feel free to resolve as such.

### How can we reproduce the bug?

I've managed to reduce it to the following test-case.

```python

async def test_anyio_cancel_shielding_start():
    entered_inner_scope = anyio.Event()
    async with (
        anyio.create_task_group() as tg,
        anyio.create_task_group() as inner_tg,
    ):
        async def inner_task(*, task_status: anyio.abc.TaskStatus = anyio.TASK_STATUS_IGNORED):
            # task_status.started() # Uncommenting this lets the task hang forever instead

            with anyio.CancelScope(shield=True):
                entered_inner_scope.set()
                try:
                    await anyio.sleep_forever()
                except anyio.get_cancelled_exc_class():
                    assert False, "Shouldn't be cancelled in a shielded scope"

        async def start_inner_task():
            await inner_tg.start(inner_task)

        tg.start_soon(start_inner_task)
        await entered_inner_scope.wait()
        tg.cancel_scope.cancel()



if __name__ == "__main__":
    anyio.run(test_anyio_cancel_shielding_start)
```

**Comment by tapetersen** [comment]
I guess the issue is the direct `asyncio.Task.cancel` here that doesn't take CancelScopes into account.

https://github.com/agronholm/anyio/blob/c967f5c644b4e535af4cb230d9bcb8be06e2559f/src/anyio/_backends/_asyncio.py#L939

We can work around it but needed to debug where the issue actually was anyway to make the workaround reliable.

**Comment by agronholm** [comment]
If you were led to understand that AnyIO cancel scopes shield from native asyncio cancellation then you were told wrong. It doesn't. AnyIO cancellation is _level cancellation_ while asyncio cancellation is _edge cancellation_. Edge cancellation does not work together with level cancellation. AnyIO cancel scopes only shield from AnyIO cancellation on asyncio. On Trio, they shield from both native and AnyIO cancellation, as both use level cancellation.

**Comment by agronholm** [comment]
I think I may have spoken too soon, and you are well versed in level cancellation semantics. Apologies. I'll try the example and see what's what.

**Comment by tapetersen** [comment]
No problem and I know this is really tricky. I just took a look through the test-cases in a "how hard can it be excercise" knowing full-well that I could expect it to be *_very_* hard. 

 I wouldn't expect the scopes to shield from external cancellation I've started myself but the confusing issue issue here is that `TaskGroup.start` uses a native cancellation if itself is cancelled using only anyio API's.
 
That you even support the case where the actual start call is natively cancelled is further than I would've expected and complicates the ideas, at least I had to maybe correct it.

**Comment by agronholm** [comment]
The example does not stumble onto the assertion error on Trio, so there is a bug of some sort. Hard to say at this point how difficult fixing this will be.

**Comment by tapetersen** [comment]
I get that and wouldn't have expected it to error on trio as I would've assumed `TaskGroup.start` is just the native method there (and can rely on all tasks having proper `CancelScope`s) .

As said before I don't expect a fast resolution here but needed to dig this out to know what workaround was needed.

The tricky part I see here is that it needs to support (and tries to from what I see in the test-cases such as test_start_native_host_cancelled) the case where the caller of `TaskGroup.start` isn't itself in a `CancelScope` and therefore can't rely on the new task being automatically cancelled by the same scope that threw the `CancelledError` we just caught (and therefore resorts to using a native `task.cancel()` call)

**Comment by agronholm** [comment]
Flat out removing the call makes the new test in the associated PR pass but others hang.

**Comment by tapetersen** [comment]
Yes, and I first tried some variations with more checks to try doing it only in the case of not coming from an anyio task (checking if we actually have a cancel_scope).

That would maybe solve it in the case of only using anyio tasks like my example while leaving the issue if you call start from a non-anyio managed task but even then it caused some issues.

Looking into the alternative of creating a new `CancelScope` for start that could be cancelled instead of using the native cancel directly (not sure if that would've been acceptable performance wise) I realised it may be tricky how that would behave with parent/child scope relations after `task_status.started` was called. 

Maybe it could work to create the new scope in the  `TaskGroup` directly even if we haven't reparented the task there yet but I don't grasp those part well enough yet to know.

Like I said, we can work around it knowing the problem  and I'm mostly poking around using it as an excuse to dig into the nitty-gritty details of the code-base even if it doesn't lead to a fix.

**Comment by agronholm** [comment]
I'm sure I'll find a workable solution in due time. Good to hear that you can at least work around the problem!

**Comment by tapetersen** [comment]
Actually solved by https://github.com/agronholm/anyio/pull/1153
